### PR TITLE
Removing dependency on RewardManager and moving many methods out of Callback

### PR DESCRIPTION
**Pull request by jdchang1**
This PR moves all the functionality from `start_iteration` callback into the `DistributedGPUActor` group. This involves the following
* `MinibatchRolloutBuffer`, `Reference Model`, and `KL Controller` are all in `DistributedGPUActor`
* `RewardManager` is now `RewardActor` which is a separate Ray actor initialized by the single controller
*  `add_rollouts` is now `create_online_minibatches` which does log prob calculation, reference and kl computation, and finally advantage calculation before creating minibatches for online training.

### CODE REVIEW

File: test_single_controller_ppo.py [hunk 1]
```diff
@@ -224,15 +331,401 @@ def build_ppo_trainer(self):
...
+            partial_batch = self.get_log_probs_and_entropy(current_rank_rollouts, device)
+            # 2) Compute Reference Log Probs and KL
+            reference_output = self.get_reference_log_probs_and_kl(partial_batch)
+ 
+            # Log to callback for KL Controller Update
+            mean_ift = masked_mean(
+                reference_output['kl'],
+                partial_batch['action_mask'],
+            )
+            self.ppo_callback.kl_ift.append(mean_ift.cpu())
```
Comment (abaheti95) [inline test_single_controller_ppo.py:363]: We shouldn't need to save the kl_ift in callback anymore since you moved it out into the distributed actor right?

File: test_single_controller_ppo.py [hunk 2]
```diff
@@ -224,15 +331,401 @@ def build_ppo_trainer(self):
...
+                **reward_output,
+                **advantage_output,
+            }
+
+            # Moving minibatches to CPU to not take additional GPU memory
+            for k, v in batch.items():
+                if hasattr(v, 'cpu'):
+                    batch[k] = v.cpu()
+
+        self.ppo_callback.batch_rollouts = batch
```
Comment (abaheti95) [inline test_single_controller_ppo.py:386]: You may have to port over the [update_ift_kl() code](https://github.com/databricks/compose-rl/blob/main/compose_rl/algorithms/online/callback.py#L1083-L1095) to after creating the batch.

File: test_single_controller_ppo.py [hunk 3]
```diff
@@ -224,15 +331,401 @@ def build_ppo_trainer(self):
...
+            mean_ift = masked_mean(
+                reference_output['kl'],
+                partial_batch['action_mask'],
+            )
+            self.ppo_callback.kl_ift.append(mean_ift.cpu())
+
+            # 3) Scale rewards and apply KL Penalty
+            reward_output = self.update_rewards(current_rank_rollouts['all_rewards_dict'], reference_output, partial_batch['action_mask'], device)
+            # 4) Compute Advantages
+            advantage_output = self.compute_advantages(partial_batch, reward_output)
```
Comment (abaheti95) [inline test_single_controller_ppo.py:368]: Maybe add a future `TODO:` that advantage calculation should be done w/ all-gather and scatter across train actor group for maximal correctness.

File: test_single_controller_ppo.py [hunk 4]
```diff
@@ -605,7 +1010,7 @@ def __init__(self, config: Any):
             rank=0,
             world_size=1,
             master_addr=None,
-            master_port=None,
+master_port=None,
```
Comment (abaheti95) [inline test_single_controller_ppo.py:1013]: bad indentation?

File: test_single_controller_ppo.py [hunk 5]
```diff
@@ -681,6 +1089,810 @@ def load_dataloader_state_dict(self, state_dict: dict):
...
+        self.pool = None
+        if self.inference_rewards or self.functional_rewards:
+            self.pool = Pool(
+                processes=len(self.inference_rewards) +
+                len(self.functional_rewards),
+                context=get_context('spawn'),
+            )
+    
+    @property
+    def fsdp_config(self):
```
Comment (abaheti95) [inline test_single_controller_ppo.py:1234]: redundant in the new design

File: test_single_controller_ppo.py [hunk 6]
```diff
@@ -681,6 +1089,810 @@ def load_dataloader_state_dict(self, state_dict: dict):
...
+        # TODO (infra): This should be the non_train_fsdp_config from the callback
+        return {}
+    
+    def initialize_composer_model(
```
Comment (abaheti95) [inline test_single_controller_ppo.py:1238]: redundant in the new design

File: test_single_controller_ppo.py [hunk 7]
```diff
@@ -681,6 +1089,810 @@ def load_dataloader_state_dict(self, state_dict: dict):
...
+                    base_batch['generated_lens'],
+                'seq_lens':
+                    base_batch['seq_lens'],
+            }
+        else:
+            raise TypeError(
+                f'Unknown reward model type {type(reward_model)}. Expected `Reward` or `RewardModel`.',
+            )
+
+    def resolve_outputs(
```
Comment (abaheti95) [inline test_single_controller_ppo.py:1549]: redundant in the new design

**Review by abaheti95 — COMMENTED** [review]
Awesome dude! This is so much cleaner!

File: test_single_controller_ppo.py [hunk 8]
```diff
@@ -158,9 +224,49 @@ def init_composer_dist(self):
...
+        # Create a Trainer object to load from checkpoint and FSDP the model
+        _ = Trainer(
+            model=self.reference_model,
+            parallelism_config=parallelism_config,
+            precision=self.precision,
+            load_weights_only=True,
+            load_strict_model_weights=False,
+            load_path=load_path,
+            python_log_level='debug',
+        )
```
Comment (bowenyang008) [inline test_single_controller_ppo.py:264]: leave a TODO that we can use FSDP2 utils to FSDP the model @rithwik-db

File: test_single_controller_ppo.py [hunk 9]
```diff
@@ -681,6 +1089,810 @@ def load_dataloader_state_dict(self, state_dict: dict):
...
+        # e.g., if special formatting is applied
+        reward_seq_len = prompt_len + generated_len
+
+        # TODO: Don't think reward_manager is actually using action log probs... its likely just a tensor for shape management
+        # TODO: cleanup reward actor methods
+        dummy_log_probs = torch.zeros_like(
+            actions,
+            dtype=torch.float32,
+        )
+        all_rewards = ray.get(self.reward_actor.calculate_reward.remote(
```
[the remark this answers is not in the document]
Comment (bowenyang008) [reply test_single_controller_ppo.py:1845]: maybe worth using the time_it context

File: yamls/single-controller-grpo-workflow.yaml [hunk 10]
```diff
@@ -145,6 +156,8 @@ parameters:
     epoch_per_iteration: 1
     generations_per_prompt: 8
     num_batches_per_update: 8
+    #generations_per_prompt: 2
+    #num_batches_per_update: 2
```
Comment (bowenyang008) [inline yamls/single-controller-grpo-workflow.yaml:160]: rm?

**Review by bowenyang008 — APPROVED** [review]
Left a couple comments, awesome works @jdchang1 !

File: yamls/single-controller-grpo-workflow.yaml [hunk 11]
```diff
@@ -137,6 +138,16 @@ parameters:
...
+    # The non-train-FSDP config
+    non_train_fsdp_config:
+      sharding_strategy: SHARD_GRAD_OP
+      mixed_precision: DEFAULT
+      activation_checkpointing: true
+      activation_cpu_offload: false
+      verbose: true
+      limit_all_gathers: true
+      state_dict_type: sharded
+      use_orig_params: true
```
[the remark this answers is not in the document]
Comment (jdchang1) [reply yamls/single-controller-grpo-workflow.yaml:150]: good question. I am not 100% sure about this. I figured we may want different fsdp configs but happy to remove if its unnecessary
